Restrict project access by IP address

esoergel opened the pull request:
## Product Description


## Technical Summary
https://dimagi.atlassian.net/browse/USH-5818
https://dimagi.atlassian.net/browse/USH-5817
[Design document](https://docs.google.com/document/d/1T2ohWAno8r3IHZZs2cvp99_VWKS7qFK_zxTXK2WxnOs/edit?tab=t.0#heading=h.ipgou4bu0qoj)

## Feature Flag


## Safety Assurance


### Safety story


### Automated test coverage


### QA Plan


### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

esoergel committed:
Add ip access model

esoergel commented:
@minhaminha here's a branch with the basic model for the IP access work - figure we can PR our respective changes into this and hopefully avoid too much conflict

esoergel committed:
add feature flag

esoergel committed:
add to dump and deletion files

esoergel committed:
Add middlware class to check IP access

esoergel committed:
Merge pull request #36072 from dimagi/es/ip-blocking-middleware

Add middlware class to check IP access

minhaminha commented on corehq/apps/ip_access/migrations/0001_initial.py:
```diff
@@ -0,0 +1,56 @@
+                    ),
+                ),
+                (
+                    "ip_denylist",
```
is there any need for a `country_denylist`?

esoergel replied:
I left it out for now since we don't need it and it makes the logic a bit harder to reason about.  Though thinking about it again I think we could make it work by saying specificity wins (ie an IP rule beats a country rule), something like:

```python
def is_allowed(ip):
    if ip in ip_denylist:
        return False
    if ip in ip_allowlist:
        return True
    if country_denylist and is_in_country(ip, country_denylist):
        return False
    return not country_allowlist or is_in_country(ip, country_allowlist):
```

That is, allowlisted IPs are always allowed even from denylisted countries.

Still, I think it's okay to hold off for now - we could always add it later if needed

minhaminha committed:
Add geoip2 to requirements files

minhaminha committed:
Use MaxMind to check if ip_address is in the country_allowlist

minhaminha commented:
@esoergel I noticed that currently, a user that's not in an allowed country according to the domain's config can trigger multiple requests if they repeatedly try to access the domain. I'm thinking we should also add a session blocklist, so that when a user fails `config.is_allowed` they aren't allowed to try again but wanted to get your thoughts as well.

esoergel commented:
@minhaminha yeah that's a good thought - I had the same idea but was uncertain whether it'd be worth it.  It certainly makes sense to cache the lookups we allow, since we _expect_ users to make a bunch of requests in sequence.  However, if their request is blocked, presumably they won't try again and again unless it's like a DOS attack, which this isn't intended to mitigate against.

Still though, I guess we wouldn't actually stick anything in the session for blocked IPs unless we _do_ encounter one, so the cost should be trivial.  I'd be on board with that.

minhaminha committed:
Add IP to session block list if is not allowed

minhaminha committed:
Fix typo, lint

minhaminha committed:
Track number of ip country lookups

minhaminha commented:
@esoergel blocking by the IP's origin country is now working properly on staging 🎉 . Is there anything else to be done on this PR before releasing it from the drafts? I created an IP Access monitoring dashboard on datadog [here](https://app.datadoghq.com/dashboard/d5j-ybm-ic9?fromUser=true&refresh_mode=sliding&from_ts=1743106990172&to_ts=1743193390172&live=true), and I'll add the requests-per-day counter on their too/setup the 1000 request threshold alert after this gets merged.  
I saw in the design doc we were considering checking IPs for API requests too - did we decide which of 3 approaches you described to go with? 
I'm also wondering if this line from the doc is necessary:

> On subsequent requests, a middleware class should check the request IP against the IP address used to initiate the session.  If the IP address differs, we flush the session and force the user to log in again.

esoergel committed:
Claim requirement for USH

esoergel commented on corehq/apps/ip_access/models.py:
```diff
@@ -0,0 +1,42 @@
+
+
+def is_in_country(ip_address, country_allowlist):
+    # If there is no countries in the allowlist, or MaxMind is not configured in the env
+    # assume all countries are allowed
```
And also probably `notify_exception`.  I think it'd also make sense to check that MaxMind is configured when displaying the config page, maybe disabling the country input if not.

esoergel committed:
Make block key take precedence over allow key

Though I can't imagine a scenario where this would matter

minhaminha committed:
Merge branch 'master' into es/ip-blocking

minhaminha committed:
Add IP access configuration page to the project settings

esoergel commented on corehq/apps/domain/views/settings.py:
```diff
@@ -271,6 +273,71 @@ def post(self, request, *args, **kwargs):
+    urlname = 'ip_access_config'
+    page_title = gettext_lazy("IP Access")
+
+    @method_decorator(always_allow_project_access)
+    @method_decorator(login_and_domain_required)
+    @use_bootstrap5
+    def dispatch(self, *args, **kwargs):
+        if IP_ACCESS_CONTROLS.enabled(self.domain):
+            return super().dispatch(*args, **kwargs)
+        else:
+            raise Http404()
```
Nit: I'd replace all this with a class-level decorator like:
```python
@method_decorator([
    always_allow_project_access,
    login_and_domain_required,
    use_bootstrap5,
    toggles.IP_ACCESS_CONTROLS.required_decorator(),
], name='dispatch')
class EditIPAccessConfigView(BaseProjectSettingsView):
```
No need to override `dispatch`

esoergel replied:
Also what's the `always_allow_project_access` decorator doing here?  I think that lets this page be accessed when the subscription is paused, but I see it's pretty widely used.

And I think this needs `domain_admin_required` instead of `login_and_domain_required`

esoergel commented on corehq/apps/domain/views/settings.py:
```diff
@@ -271,6 +273,71 @@ def post(self, request, *args, **kwargs):
+
+            if should_save:
+                domain_config.save()
+                messages.success(request, _("Your IP Access settings have been saved!"))
```
This should redirect on success so reloading the page doesn't resubmit the form

esoergel commented on corehq/apps/domain/views/settings.py:
```diff
@@ -271,6 +273,71 @@ def post(self, request, *args, **kwargs):
+            return IPAccessConfigForm(self.request.POST, initial=initial)
+        return IPAccessConfigForm(initial=initial)
+
+    @property
+    def get_ip_access_config(self):
```
Nit: I'd either make this a `@cached_property` and rename it `ip_access_config` or drop `@property`.  It's surprising to see something called `get_*` that you don't call with parenthesis.

esoergel commented on corehq/apps/domain/views/settings.py:
```diff
@@ -271,6 +273,71 @@ def post(self, request, *args, **kwargs):
+                domain_config.domain = self.domain
+
+            for attr, value in self.form.cleaned_data.items():
+                if value:
```
```suggestion
                if value != getattr(domain_config, attr):
```
Without this, the user won't be able to blank out any fields.  For instance, if they have a comment on the saved `domain_config`, then submit the form to delete that comment, `value` will be falsy and this won't update.

esoergel commented on corehq/apps/domain/forms.py:
```diff
@@ -222,6 +223,79 @@ def save(self, user, domain):
+
+        # Ensure an IP isn't in both lists
+        if (allow_list or deny_list) and set(allow_list).intersection(set(deny_list)):
+            raise ValidationError(_("There are IP addresses in both the Allowed and Blacklisted lists. "
+                                    "Please ensure an IP address is only in one list at a time."))
```
:+1:

esoergel commented on corehq/apps/domain/forms.py:
```diff
@@ -222,6 +223,79 @@ def save(self, user, domain):
+    )
+
+    ip_denylist = forms.CharField(
+        label="Blacklisted IPs",
```
why "Blacklisted"?  I think of it as either we call them whitelist and blacklist or allowlist and denylist (and I think the latter pair is more in favor).

minhaminha committed:
Various PR feedback changes

esoergel commented on corehq/apps/domain/views/settings.py:
```diff
@@ -273,25 +273,24 @@ def post(self, request, *args, **kwargs):
-            return super().dispatch(*args, **kwargs)
-        else:
-            raise Http404()
+        return super().dispatch(*args, **kwargs)
```
Nit: you can now delete this method

esoergel reviewed (commented):
No blocking feedback, though it looks like I can't approve the PR since I opened it

minhaminha committed:
Remove unncessary dispatch override method

minhaminha commented:
> No blocking feedback, though it looks like I can't approve the PR since I opened it

I can't either as the last pusher but maybe a good idea to have someone else review it anyways

MartinRiese commented:
Got half way through the commits so far.

Have you thought about supporting subnet masks like `192.168.1.0/24`? Especially state governments in the US probably have a good block to themselves and would rather not write out all the allowed ips.

Jtang-1 commented on corehq/apps/domain/forms.py:
```diff
@@ -222,6 +223,79 @@ def save(self, user, domain):
+    """
+    country_allowlist = forms.MultipleChoiceField(
+        label="Allowed Countries",
+        choices=sorted(list(COUNTRIES.items()), key=lambda x: x[0]),
```
Should this be sorted by `x[1]` so that it's sorted by the human readable text

Jtang-1 commented on corehq/apps/domain/forms.py:
```diff
@@ -222,6 +223,79 @@ def save(self, user, domain):
+    )
+
+    ip_allowlist = forms.CharField(
+        label="Allowed IPs",
```
this should be translated

Jtang-1 commented on corehq/apps/domain/forms.py:
```diff
@@ -222,6 +223,79 @@ def save(self, user, domain):
+        deny_list = self.cleaned_data['ip_denylist'].split(", ") if self.cleaned_data['ip_denylist'] else []
+
+        # Ensure an IP isn't in both lists
+        if (allow_list or deny_list) and set(allow_list).intersection(set(deny_list)):
```
nit: I think this can be `allow_list and deny_list`

Jtang-1 commented on corehq/apps/domain/views/settings.py:
```diff
@@ -271,6 +273,71 @@ def post(self, request, *args, **kwargs):
+            'ip_access_config_form': self.form,
+        }
+
+    def post(self, request, *args, **kwargs):
```
nit: suggestion for improved readability. Feel free to disagree and not take any or take parts of the suggestion below

```
def post(self, request, *args, **kwargs):
    if not self.form.is_valid():
        return self.get(request, *args, **kwargs)

    domain_config = self.ip_access_config or IPAccessConfig(domain=self.domain)
    should_save = False

    for attr, value in self.form.cleaned_data.items():
        current_value = getattr(domain_config, attr, None)
        if value != current_value:
            setattr(domain_config, attr, value)
            should_save = True

    if should_save:
        domain_config.save()
        messages.success(request, _("Your IP Access settings have been saved!"))
        return HttpResponseRedirect(reverse(self.urlname, args=[self.domain]))

    return self.get(request, *args, **kwargs)
 ```

minhaminha replied:
I have the `if not domain_config` conditional block in case for whatever reason a new domain submits an empty form and to prevent the form from creating and saving a new config object with nothing in it

esoergel commented:
> Have you thought about supporting subnet masks like 192.168.1.0/24? Especially state governments in the US probably have a good block to themselves and would rather not write out all the allowed ips.

@MartinRiese that sounds like a good idea - looks like the https://docs.python.org/3/library/ipaddress.html python library should make that reasonably straightforward.  Still, I'd maybe push that to a followup PR

esoergel commented on corehq/apps/ip_access/middleware.py:
```diff
@@ -0,0 +1,48 @@
+    if allow_key not in request.session:
+        request.session[allow_key] = []
+
+    if ip in request.session[block_key]:
```
I've been working on the assumption that this would be used primarily on domains with the "shorten inactivity timeout" setting enabled, which expires sessions [after 30 minutes of inactivity](https://commcare-hq.readthedocs.io/middleware.html#timeoutmiddleware).  Though session expiration isn't actually covered in the tests here - I'll see if I can add that.

esoergel committed:
Add test for session invalidation

minhaminha committed:
Small changes based on PR feedback

minhaminha commented on corehq/apps/ip_access/models.py:
```diff
@@ -0,0 +1,42 @@
+def is_in_country(ip_address, country_allowlist):
+    # If there is no countries in the allowlist, or MaxMind is not configured in the env
+    # assume all countries are allowed
+    if not country_allowlist or not settings.MAXMIND_LICENSE_KEY:
```
Can you say more about why this shouldn't handle an empty `country_allowlist`?

Also I didn't really think about it until now but what happens when a third party hoster without a license key tries to set up the configuration and is blocked from accessing their domain? Is it reasonable to tell them to change it manually in shell? cc @esoergel

Jtang-1 replied:
I take back what I said. It makes sense to handle in the function, but I would expect the function to return `False` if the `country_allowlist` is empty, since the ip_address wouldn't be in the allowlist. I wouldn't expect it to return `True`.

Jtang-1 replied:
> On the flip side if only the ip_allowlist is populated and an IP isn't on that list, we probably shouldn't allow it just because the country list is empty either.

Agreed. 

There are two aspects to consider: the UI and the code.

UI:
Currently, an empty `ip_allowlist` implies "full restriction," while an empty `country_allowlist` means "no restriction." This could be confusing. For better clarity, we should display default text such as "All countries" when no countries are selected.

Code:
To make the logic clearer, we could introduce an `allow_all_countries` flag. This way, when the form is submitted with no value for `country_allowlist`, we would set `country_allowlist` to an empty list ([]) and set `allow_all_countries` to True. I think this aligns with other parts of our codebase where we handle similar defaults.

Example implementation:

```
    def is_allowed(self, ip_address):
        return (
            ip_address not in self.ip_denylist
            and (
                ip_address in self.ip_allowlist
                or self.country_allowed()
            )
        )
  
    @property
    def country_allowed(self):
        return self.allow_all_countries or is_in_country(ip_address, self.country_allowlist)
```

(interesting forum [post](https://ux.stackexchange.com/questions/36582/does-leaving-a-field-blank-mean-everyone) that talks about this)

esoergel committed:
Rejigger no_permissions handler and log

This lets us use custom exception classes to pass context to the page
displayed to the user, while using the same code path for permissions
errors raised and permissions error responses

esoergel committed:
Return 403 instead of 451 for bad IP addresses

esoergel committed:
Merge pull request #36198 from dimagi/es/ip-access-error

Return HTTP 403 on bad IP

minhaminha committed:
Add more guards to prevent users from blocking themselves out of the project

minhaminha commented:
Sorry for the delay yall I kept confusing myself trying to think of every scenario in which a user could block themselves from their own project but I think the last commit covers everything. The only thing I didn't do was throw hard errors if the env doesn't have a license key (it'll still block country configuration if there isn't any) but instead had it show an error banner saying they need to do that, since maybe they could still use for the allow and deny lists.

minhaminha committed:
Text fixes + reinstating is_in_country

minhaminha committed:
Another attempt at fixing tests + slight refactor

Jtang-1 commented on corehq/apps/ip_access/models.py:
```diff
@@ -0,0 +1,46 @@
+
+    def is_allowed(self, ip_address):
+        should_check_country = True
+        if not self.country_allowlist:
```
Don't we only need to check Maxmind license only if `country_allowlist` is populated? 

Should the logic be (I'm assuming an empty self.country_allowlist means allow no countries)


```
def is_allowed(self, ip_address):
    if ip_address in self.ip_denylist:
        return False

    if ip_address in self.ip_allowlist:
        return True

    if self.country_allowlist and not settings.MAXMIND_LICENSE_KEY:
        return False

    return is_in_country(ip_address, self.country_allowlist)

```

esoergel replied:
Yeah doesn't this change it so `country_allowlist = []` means no countries are allowed?

Jtang-1 commented on corehq/apps/domain/views/settings.py:
```diff
@@ -322,17 +332,41 @@ def post(self, request, *args, **kwargs):
+                messages.error(request,
+                               _("You must provide either your country or your "
+                                 "current IP Address to have access to this project."))
+            else:
```
nice checks for configurations that would block themselves

esoergel commented on corehq/apps/domain/views/settings.py:
```diff
@@ -271,6 +273,103 @@ def post(self, request, *args, **kwargs):
+                should_save = False
+                domain_config = IPAccessConfig()
+                domain_config.domain = self.domain
+            if not self.form.cleaned_data["ip_allowlist"] and not self.form.cleaned_data["country_allowlist"]:
+                should_save = False
+                messages.error(request,
+                               _("You must provide either your country or your "
+                                 "current IP Address to have access to this project."))
```
I don't follow this clause - is it saying that one of the two must be provided?  If a user specifies only an `ip_denylist`, what's the problem?

esoergel commented on corehq/apps/domain/views/settings.py:
```diff
@@ -271,6 +273,103 @@ def post(self, request, *args, **kwargs):
+        }
+
+    def post(self, request, *args, **kwargs):
+        if self.form.is_valid():
```
Can you move some/all of the below validation checks into the form, so validation is all in one place?  That'd also let errors be attached to the associated fields automatically.

esoergel replied:
It wouldn't give as detailed a message, but you could also just do a generic "will this config let me access the project" check like so:
```python
IPAccessConfig(
    country_allowlist=form.cleaned_data['country_allowlist'],
    ip_allowlist=form.cleaned_data['ip_allowlist'],
    ip_denylist=form.cleaned_data['ip_denylist'],
).is_allowed(get_ip(request))
```
That would ensure the validation matches the implementation exactly

esoergel commented on corehq/apps/domain/views/settings.py:
```diff
@@ -271,6 +273,103 @@ def post(self, request, *args, **kwargs):
+            else:
+                for attr, value in self.form.cleaned_data.items():
+                    current_value = getattr(domain_config, attr)
+                    if value != current_value:
```
This now means it only checks the country allowlist if it changes, and since validation happens inside here, it won't always be called.  You could set up a config which blocks your country but allows your IP, then on a subsequent request, delete your IP from the allowlist, causing you to lose access.

minhaminha replied:
oh good catch

esoergel committed:
Add tests for various types of country configs

minhaminha committed:
Refactor so an empty country_allowlist allows all countries + moving validation steps into form cleaning + raise 403 if license key isn't valid + test fixes

Jtang-1 commented on corehq/apps/ip_access/middleware.py:
```diff
@@ -46,9 +46,9 @@ def is_valid_ip(request, domain):
         config = IPAccessConfig.objects.get(domain=domain)
     except IPAccessConfig.DoesNotExist:
         config = None
+    if config and not settings.MAXMIND_LICENSE_KEY:
```
should this be `if config and config.country_allowlist and not settings.MAXMIND_LICENSE_KEY`

minhaminha replied:
`config.country_allowlist and not settings.MAXMIND_LICENSE_KEY` will be handled by `get_ip_country` down the line that'll raise a 403 error saying basically the same message as this, but as a hard error. This is more the non blocking/gentle reminder error message to set up MaxMind / deter forcibly assigning country_allowlist in shell or something like that.

Jtang-1 replied:
Wouldn’t this message always show up, even for domains that don’t use the country_allowlist config? My understanding is that the MaxMind key is only required if the IP access config uses country_allowlist. If that’s the case, users who aren’t using it wouldn’t need the key and wouldn’t expect to see this message.

Although, if the domain is using country_allowlist without the MaxMind key, wouldn’t the request already fail with a 500 in `is_allowed`? In that case, the user wouldn’t even reach this page to see the error, so the message might not be necessary at all.

minhaminha replied:
yea I wasn't sure if we wanted to explicitly allow this feature to be used just for the IP allow and deny lists, so the point of this was to remind users (only upon entering a new session) that having a MaxMind license is something we strongly suggest, but I guess there's not really strong reasons to do that. I can remove this altogether and rely solely on the 500 error thrown in `is_allowed` which will then only remind users of the missing license key when they're trying to use the country filtering feature.

Jtang-1 reviewed (commented):
just the one comment, otherwise looks good to me!

esoergel commented on corehq/apps/ip_access/models.py:
```diff
@@ -25,27 +28,24 @@ def __repr__(self):
-                or (should_check_country and is_in_country(ip_address, self.country_allowlist))
-            )
-        )
+        if ip_address in self.ip_denylist or (self.country_allowlist and not settings.MAXMIND_LICENSE_KEY):
+            return False
```
I think you can drop the `or..` part of this conditional.  As-is, this means it won't call `is_in_country`, and the 403 will never be raised.

esoergel commented on corehq/apps/ip_access/models.py:
```diff
@@ -25,27 +28,24 @@ def __repr__(self):
+            metrics_counter('commcare.ip_access.check_country')
+            return response.country.iso_code
+    except geoip2.errors.AuthenticationError:
+        raise Http403(_("Please provide a valid MaxMind license key and Account ID to filter IPs by country. "
+                      "If you do not have a MaxMind account, please clear the Allowed Countries field in the "
+                      "IP Access Config to access your project."))
```
Actually this also needs to be a 500 error or it won't show up in sentry (or other error logging) at all, which would make this hard to debug.

minhaminha committed:
Minor refactor + test fix

minhaminha committed:
Use notify_exception instead of raising 403 + remove unncessary error banner

minhaminha committed:
Add more tests

esoergel committed:
Allow empty values in django admin

minhaminha committed:
Delete print statement + raise Exception instead only notifying it

minhaminha committed:
Test fix

esoergel commented:
Just did another full pass - I have no open feedback, so I think we're g2g from my end

Jtang-1 reviewed (approved):
looks good to me
